lib/cartage/plugins/remote.rb

```diff
@@ -0,0 +1,481 @@
+
+    def initialize(*) #:nodoc:
+      super
+      @host = @name = @keys = @key_data = nil
+
+      @stage = MicroMachine.new('new').tap { |stage|
+        stage.when :local_setup, 'new' => 'local_config'
+        stage.when :ssh_setup, 'local_config' => 'ssh_config'
+        stage.when :run_prebuild, 'ssh_config' => 'prebuild'
+        stage.when :clone_remote, 'prebuild' => 'remote_clone'
+        stage.when :build_remote, 'remote_clone' => 'remote_build'
+        stage.when :clean_remote, 'remote_build' => 'cleanup'
+        stage.when :complete, 'cleanup' => 'finished'
+        stage.on(:any) { |event| dispatch(event, stage.state) }
+      }
+    end
+
+    attr_reader :host
+    attr_reader :name
+    attr_reader :config
+    attr_reader :stage
+
+    attr_reader :build_root
+    attr_reader :config_file
+    attr_reader :key_data
+    attr_reader :keys
+    attr_reader :paths
+    attr_reader :subs
+
+    def prebuild_script
+      unless defined?(@prebuild_script)
+        @prebuild_script = host.prebuild || config.prebuild ||
+          DEFAULT_PREBUILD_SCRIPT
+      end
+      @prebuild_script
+    end
+
+    def postbuild_script
+      unless defined?(@postbuild_script)
+        @postbuild_script = host.postbuild || config.postbuild
+      end
+      @postbuild_script
+    end
+
+    def build_script
+      unless defined?(@build_script)
+        @build_script = host.build || config.build || DEFAULT_BUILD_SCRIPT
+      end
+      @build_script
+    end
+
+    def tmpfiles
+      @tmpfiles ||= []
+    end
+
+    def local_setup
+      @name = config.host || 'default'
+      host_config = config.dig(:hosts, name)
+      verify_host!(name, host_config)
+      @host = Cartage::Remote::Host.new(host_config)
+
+      fail ArgumentError, <<-exception if build_script.nil? || build_script.empty?
+No build script to run on remote #{host}.
+      exception
+
+      # Force lazy values to be present during execution.
+      cartage.send(:realize!)
+
+      cartage.display 'Pre-build configuration...'
+      @paths = OpenStruct.new(build_root: build_root)
+      paths.cartage_path = paths.build_root.join('cartage')
+      paths.project_path = paths.cartage_path.join(cartage.name)
+      paths.isolation_path = paths.project_path.join(cartage.timestamp)
+      paths.build_path = paths.isolation_path.join(cartage.name)
+      paths.remote_bundle = paths.isolation_path.join('deps')
+      paths.dependency_cache = paths.project_path
+      paths.config_file = paths.isolation_path.join('cartage.yml')
+      paths.build_script = paths.isolation_path.join('cartage-build-remote')
+
+      @config_file = make_config(paths).path
+
+      @subs = OpenStruct.new(
+        paths.to_h.merge(
+          repo_url:        cartage.repo_url,
+          name:            cartage.name,
+          release_hashref: cartage.release_hashref,
+          timestamp:       cartage.timestamp,
+          remote_host:     host.address,
+          remote_address:  host.address,
```

Comment:
That‘s actually intentional for backwards compatibility with older scripts. The proper name is `remote_address`, but older scripts may have `remote_host`.